File: LeetCode/1. Two Sum/Solution.kt
package problem1

class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val len = nums.size
        for (i in 0 until len) {
            for (j in 0 until len) {
                if (i != j && nums[i] + nums[j] == target) {
                    return intArrayOf(i, j)
                }
            }
        }
        return intArrayOf(0, 0)
    }
}

fun test(
    // input
    nums: IntArray, target: Int,
    // expected
    expected: IntArray,
) {

    val result = Solution().twoSum(nums, target)
    println("(${nums.contentToString()}, $target) -> ${result.contentToString()} == ${expected.contentToString()} -> " +
        "${result.contentEquals(expected)}")
}

fun main() {
    test(
        intArrayOf(2, 7, 11, 15), 9,
        intArrayOf(0, 1)
    )

    test(
        intArrayOf(3, 2, 4), 6,
        intArrayOf(1, 2)
    )

    test(
        intArrayOf(3, 3), 6,
        intArrayOf(0, 1)
    )
}

File: LeetCode/1. Two Sum/Solution.cc
 #include <algorithm> // for std::min
#include <unordered_map>
 #include <vector>

using std::vector;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // value to index
        std::unordered_map<int, int> map;

        for (int i = 0; i < nums.size(); i++) {
            int x = target - nums[i];
            auto it = map.find(x);
            if (it != map.end()) {
                return std::vector<int>({it->second, i});
            }
            map[nums[i]] = i;
        }

        return std::vector<int>({0, 0});
    }
};

// test

#include <iostream>

int main() {
}